main/src/main/scala/sbt/internal/server/LanguageServerProtocol.scala

```diff
@@ -135,6 +136,47 @@ private[sbt] object LanguageServerProtocol {
                 case NonFatal(e) =>
                   errorRespond("Cancel request failed")
               }
+            case r: JsonRpcRequestMessage if r.method == "textDocument/completion" =>
```

Comment:
Is this to complete commands for the sbt shell of complete Scala source code? `textDocument/completion` is meant for completion in text documents. This is this for sbt shell completions then I think it makes sense to use a different method name like `sbt/completion`. It can still use the LSP data structures like `CompletionContext` https://microsoft.github.io/language-server-protocol/specification#textDocument_completion

Reply:
Very cool feature btw! 😄

Reply:
@olafurpg correct, this is for sbt interactive shell completions, I don't mind much on the `method` name.. :smile:
Sbt completions are kind of complicated since they flow down to user-code class names for example (i.e. `testOnly`).
I leave the final word to @eed3si9n

Reply:
On Gitter chat I wrote:

> eugene yokota eed3si9n Sep 28 05:29
> CompletionParams extends TextDocumentPositionParams
so in the request it's supposed to containt the position of which file the user's on and where
since we are using for shell purpose, I think we need to make up a non-existent URL for the shell
>  I guess textDocument: TextDocumentIdentifier; is but I don't want to assume it's for shell
since normally it's used for Scala source's completion

I think the editor interested in sbt shell completion could still use "textDocument/completion" but it should pass the URL `sbt://shell`, and sbt server should respond only for that.

Reply:
I tend to agree on @olafurpg point of view, makes totally sense.

Reply:
Yes. These are actually good points. Let's do sbt specific completion command.